src/azure-cli-core/azure/cli/core/azclierror.py

```diff
@@ -4,85 +4,178 @@
         from azure.cli.core.azlogging import CommandLoggerContext
         with CommandLoggerContext(logger):
-            message = '{}: {}'.format(self.error_type.value, self.error_msg)
+            # print error type and error message
+            message = '{}: {}'.format(self.__class__.__name__, self.error_msg)
             logger.error(message)
-            if self.raw_exception:
-                logger.exception(self.raw_exception)
+            # print exception trace if there is
+            if self.exception_trace:
+                logger.exception(self.exception_trace)
+            # print recommendations to action
             if self.recommendations:
                 for recommendation in self.recommendations:
                     print(recommendation, file=sys.stderr)
 
     def send_telemetry(self):
-        import azure.cli.core.telemetry as telemetry
-        telemetry.set_error_type(self.error_type.value)
-
-        # For userfaults
-        if self.error_type in [AzCLIErrorType.CommandNotFoundError,
-                               AzCLIErrorType.ArgumentParseError,
-                               AzCLIErrorType.ValidationError,
-                               AzCLIErrorType.ManualInterrupt]:
-            telemetry.set_user_fault(self.error_msg)
-
-        # For failures: service side error, client side error, unexpected error
-        else:
-            telemetry.set_failure(self.error_msg)
-
-        # For unexpected error
-        if self.raw_exception:
-            telemetry.set_exception(self.raw_exception, '')
+        telemetry.set_error_type(self.__class__.__name__)
+# endregion
+
+
+# region: Second Layer
+# Main categories of the AzureCLI error types, used for Telemetry alalysis
+# DO NOT raise the error classes here directly in your codes.
+class UserFault(AzCLIError):
+    """ Users should be responsible for the errors. """
+    def send_telemetry(self):
+        super().send_telemetry()
+        telemetry.set_user_fault(self.error_msg)
+
+
+class ServiceError(AzCLIError):
+    """ Azure Services should be responsible for the errors.  """
+    def send_telemetry(self):
+        super().send_telemetry()
+        telemetry.set_failure(self.error_msg)
+
+
+class ClientError(AzCLIError):
+    """ Azure CLI should be responsible for the errors. """
+    def send_telemetry(self):
+        super().send_telemetry()
+        telemetry.set_failure(self.error_msg)
+        if self.exception_trace:
+            telemetry.set_exception(self.exception_trace, '')
+# endregion
+
+
+# region: Third Layer
+# Sub-categories of the AzureCLI error types, shown to users
+# Raise the error classes here in your codes
+# Command related error types
+class CommandNotFoundError(UserFault):
+    """ Command is misspelled or not recognized by AzureCLI. """
+    pass
+
+
+# Argument related error types
+class UnrecognizedArgumentError(UserFault):
+    """ Argument is misspelled or not recognized by AzureCLI. """
+    pass
```

Comment:
Why many empty class definitions?

Reply:
We can do some customiztions if necesary later, but now i think it's enough